kernel/src/engine/default/json.rs

```diff
@@ -127,90 +179,309 @@ pub struct JsonOpener {
+        inner: T,
+        // Combined state: queue and wakers, protected by a single mutex
+        state: Mutex<KeysAndWakers>,
+    }
+
+    #[derive(Debug, Default)]
+    struct KeysAndWakers {
+        // Queue of paths in order which they will resolve
+        ordered_keys: VecDeque<Path>,
+        // Map of paths to wakers for pending get requests
+        wakers: HashMap<Path, Waker>,
+    }
+
+    impl<T: ObjectStore> OrderedGetStore<T> {
+        fn new(inner: T, ordered_keys: impl Into<VecDeque<Path>>) -> Self {
+            let ordered_keys = ordered_keys.into();
+            // Check for duplicates
+            let mut seen = HashSet::new();
+            for key in ordered_keys.iter() {
+                if !seen.insert(key) {
+                    panic!("Duplicate key in OrderedGetStore: {}", key);
+                }
+            }
+
+            let state = KeysAndWakers {
+                ordered_keys,
+                wakers: HashMap::new(),
+            };
+
+            Self {
+                inner,
+                state: Mutex::new(state),
+            }
+        }
+    }
+
+    impl<T: ObjectStore> std::fmt::Display for OrderedGetStore<T> {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            let state = self.state.lock().unwrap();
+            write!(f, "OrderedGetStore({:?})", state.ordered_keys)
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl<T: ObjectStore> ObjectStore for OrderedGetStore<T> {
+        async fn put(&self, location: &Path, payload: PutPayload) -> Result<PutResult> {
+            self.inner.put(location, payload).await
+        }
+
+        async fn put_opts(
+            &self,
+            location: &Path,
+            payload: PutPayload,
+            opts: PutOptions,
+        ) -> Result<PutResult> {
+            self.inner.put_opts(location, payload, opts).await
+        }
+
+        async fn put_multipart(&self, location: &Path) -> Result<Box<dyn MultipartUpload>> {
+            self.inner.put_multipart(location).await
+        }
+
+        async fn put_multipart_opts(
+            &self,
+            location: &Path,
+            opts: PutMultipartOpts,
+        ) -> Result<Box<dyn MultipartUpload>> {
+            self.inner.put_multipart_opts(location, opts).await
+        }
+
+        // A GET request is fulfilled by checking if the requested path is next in order:
+        // - if yes, remove the path from the queue and proceed with the GET request, then wake the
+        //   next path in order
+        // - if no, register the waker and wait
+        async fn get(&self, location: &Path) -> Result<GetResult> {
+            // Do the actual GET request first, then introduce any artificial ordering delays as needed
+            let result = self.inner.get(location).await;
+
+            // we implement a future which only resolves once the requested path is next in order
+            future::poll_fn(move |cx| {
```

Comment:
Why not just have this return `Poll::Ready(result)`?

Reply:
`poll_fn` takes an `FnMut` so it would require us being able to either (1) capture the result into the closure multiple times (impossible - would move multiple times) or (2) we would have to just directly do the `self.inner.get` inside the `poll_fn` which I think is also difficult since `poll_fn` is synchronous and we want to be able to `.await`.

let me know if I'm missing something but i played with it for a second and came up with those items!